### PR TITLE
feat(checkout-button): front-end product variation selection

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-theme/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Allow the product variation selection to be made by the reader as a pre-checkout modal.

image

image

### How to test the changes in this Pull Request:

1. Create a production with variations
2. Make sure to include a description text to at least one of the variations
3. Add a Checkout Button block to a page and select the product
4. Confirm the default state is as above: "Allow the reader to select the variation before checkout." checked and empty variation selection (`--`)
5. Save, visit the page, and confirm clicking the button renders the pre-checkout modal
6. Click on any option and confirm checkout loads with the selected variation
7. Close the modal, click on the button again, select a different variation, and confirm it loads the checkout with the selected variation
8. Edit the page, toggle off the "Allow the reader to select the variation before checkout", select a variation and save
9. Visit the page and confirm it renders the checkout for the selected variation
10. Edit the page, change the product to something else, and save
11. Visit the page and confirm it renders the checkout for the selected product

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: src/blocks/checkout-button/edit.js [hunk 1]
```diff
@@ -261,10 +269,14 @@ function CheckoutButtonEdit( props ) {
 											'newspack-blocks'
 										) }
 										value={ variation }
-										options={ variations.map( item => ( {
-											label: getVariationName( item ),
-											value: item.id,
-										} ) ) }
+										disabled={ ! variation }
+										options={ [
+											{ label: __( '--', 'newspack-blocks' ), value: '' },
```
Comment (leogermani) [inline src/blocks/checkout-button/edit.js:274]: nit: do we really need to translate `--`?

File: src/modal-checkout/modal.js [hunk 2]
```diff
@@ -54,13 +58,64 @@ domReady( () => {
...
 			form.target = iframeName;
-			form.addEventListener( 'submit', () => {
+			form.addEventListener( 'submit', ev => {
+				const formData = new FormData( form );
+				// Clear any open variation modal.
+				variationModals.forEach( variationModal => ( variationModal.style.display = 'none' ) );
+				// Trigger variation modal if variation is not selected.
+				if ( formData.get( 'is_variable' ) && ! formData.get( 'variation_id' ) ) {
+					const variationModal = [ ...variationModals ].find( modal =>
+						modal.matches( "[data-product-id='" + formData.get( 'product_id' ) + "']" )
```
Comment (leogermani) [inline src/modal-checkout/modal.js:110]: what about using `modal.dataset.product-id === formData.get('product_id')` instead of concatenating ?

**Review by leogermani — DISMISSED** [review]
Code looks good to me! Modals only printed to the footer when needed :) ]

Left two small non-blocking comments

**Comment by miguelpeixe** [comment]
Thank you, @laurelfulford and @leogermani!

The issue reported by @laurelfulford is because the attribute should be a string, a2ea488d705b79d73b23c351f8dd06efef23de6a casts the integer to string before setting the attribute. Should be resolved now!

**Review by laurelfulford — APPROVED** [review]
Thanks @miguelpeixe! 🙌 

Retested the button, and I can't recreate the original issue; I also ran through switching it back and forth from allowing readers to pick the variations, and between variation and non variation products, and all looks good to me -- giving it a thumbs up from a functional perspective!

Also I meant to add in the original review that this little UX, with the checkbox re-checking itself when you pick "--", is super nice!

product-variations

**Pull request merged**